routemaster/cron.py

```diff
@@ -2,79 +2,155 @@
 
 import time
 import threading
+import contextlib
+from typing import Callable, Iterator
 
 import schedule
 
 from routemaster.app import App
 from routemaster.config import (
     Gate,
+    State,
     Action,
     TimeTrigger,
     StateMachine,
     IntervalTrigger,
+    MetadataTrigger,
 )
+from routemaster.state_machine import process_action_retries
 
 
-def _trigger_action(action: Action) -> None:
-    print(f"Processing action {action.name}")
+def _retry_action(app, state, state_machine, process_part) -> None:
+    print(f"Processing action retries for {state.name}")
+    process_action_retries(app, state, state_machine, process_part)
 
 
-def _trigger_gate(gate: Gate) -> None:
-    print(f"Processing gate {gate.name}")
+def _trigger_gate(app, state, state_machine, process_part) -> None:
+    print(f"Processing interval/time trigger for {state.name}")
+
+
+def _retry_metadata_updates(app, state, state_machine, process_part) -> None:
+    print(f"Processing metadata update trigger for {state.name}")
 
 
 def _configure_schedule_for_state_machine(
+    app: App,
     scheduler: schedule.Scheduler,
     state_machine: StateMachine,
+    is_terminating: Callable[[], bool],
 ) -> None:
+    def _process(fn: CronProcessor, state: State) -> None:
+        return _process_cron_job(is_terminating, fn, app, state_machine, state)
+
     for state in state_machine.states:
         if isinstance(state, Action):
             scheduler.every().minute.do(
-                _trigger_action,
+                _process,
+                _retry_action,
                 state,
             )
-            pass
+
         elif isinstance(state, Gate):
             for trigger in state.triggers:
                 if isinstance(trigger, TimeTrigger):
                     scheduler.every().day.at(
                         f"{trigger.time.hour:02d}:{trigger.time.minute:02d}",
                     ).do(
+                        _process,
                         _trigger_gate,
                         state,
                     )
                 elif isinstance(trigger, IntervalTrigger):
                     scheduler.every(
-                        trigger.interval.total_seconds()
+                        trigger.interval.total_seconds(),
                     ).seconds.do(
+                        _process,
                         _trigger_gate,
                         state,
                     )
+                elif isinstance(trigger, MetadataTrigger):
+                    scheduler.every(60).seconds.do(
```

Comment:
For consistency with the action triggering, this might be clearer as `scheduler.every().minute.do(`.